examples/pbf2d.py

```diff
@@ -232,90 +215,60 @@ def compute_position_deltas():
         pos_delta_i = ti.Vector([0.0, 0.0])
         for j in range(particle_num_neighbors[p_i]):
             p_j = particle_neighbors[p_i, j]
-            # TODO: does taichi supports break?
-            if p_j >= 0:
-                lambda_j = lambdas[p_j]
-                pos_ji = pos_i - positions[p_j]
-                scorr_ij = compute_scorr(pos_ji)
-                pos_delta_i += (lambda_i + lambda_j + scorr_ij) * \
-                    spiky_gradient(pos_ji, h)
+            if p_j < 0:
+                break
+            lambda_j = lambdas[p_j]
+            pos_ji = pos_i - positions[p_j]
+            scorr_ij = compute_scorr(pos_ji)
+            pos_delta_i += (lambda_i + lambda_j + scorr_ij) * \
+                spiky_gradient(pos_ji, h)
 
         pos_delta_i /= rho0
         position_deltas[p_i] = pos_delta_i
-
-
-@ti.kernel
-def apply_position_deltas():
     for i in positions:
         positions[i] += position_deltas[i]
 
 
 @ti.kernel
-def update_velocities():
+def epilogue():
+    for i in positions:
+        pos = positions[i]
+        positions[i] = confine_position_to_boundary(pos)
     for i in positions:
         velocities[i] = (positions[i] - old_positions[i]) / time_delta
+    # no vorticity/xsph because we cannot do cross product in 2D...
 
 
 def run_pbf():
-    blit_buffers(positions, old_positions)
-    apply_gravity_within_boundary()
-
-    grid_num_particles.fill(0)
-    particle_neighbors.fill(-1)
-    update_grid()
-    find_particle_neighbors()
+    prologue()
     for _ in range(pbf_num_iters):
-        compute_lambdas()
-        compute_position_deltas()
-        apply_position_deltas()
-
-    confine_to_boundary()
-    update_velocities()
-    # no vorticity/xsph because we cannot do cross product in 2D...
```

Comment:
We can do cross 2D now, so what should I do?

Reply:
Nothing really. We need the direction, not the magnitude.